File: packages/tools/tool-agent-cli/src/run/env/resolve.ts
import { createHash } from "node:crypto";
import { existsSync, lstatSync, realpathSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import {
  DEFAULT_NIXPKGS_PIN,
  NIXPKGS_PINS,
  type EnvSpec,
  type ResolvedEnv,
} from "./types.js";

/**
 * Base directory for agent-nix data
 */
export function getAgentNixDir(): string {
  return join(homedir(), ".local", "share", "agent-nix");
}

/**
 * Directory for storing generated .nix spec files
 */
export function getSpecsDir(): string {
  return join(getAgentNixDir(), "specs");
}

/**
 * Directory for storing nix-build output symlinks
 */
export function getEnvsDir(): string {
  return join(getAgentNixDir(), "envs");
}

/**
 * Directory for per-agent runtime data
 */
export function getAgentsDir(): string {
  return join(getAgentNixDir(), "agents");
}

/**
 * Regex for validating package names
 * Allows alphanumeric, underscore, plus, hyphen, and dot
 */
const PACKAGE_NAME_REGEX = /^[\w+\-.]+$/;

/**
 * Validate a single package name
 */
export function validatePackageName(name: string): boolean {
  return PACKAGE_NAME_REGEX.test(name);
}

/**
 * Validate an EnvSpec
 * @throws Error if validation fails
 */
export function validateEnvSpec(spec: EnvSpec): void {
  // Validate nixpkgs_pin
  if (!spec.nixpkgs_pin) {
    throw new Error("nixpkgs_pin is required");
  }
  if (!(spec.nixpkgs_pin in NIXPKGS_PINS)) {
    const allowed = Object.keys(NIXPKGS_PINS).join(", ");
    throw new Error(
      `Invalid nixpkgs_pin "${spec.nixpkgs_pin}". Allowed: ${allowed}`,
    );
  }

  // Validate packages
  if (spec.packages.length === 0) {
    throw new Error("packages must be a non-empty array");
  }

  for (const pkg of spec.packages) {
    if (!validatePackageName(pkg)) {
      throw new Error(
        `Invalid package name "${pkg}". Must match ${PACKAGE_NAME_REGEX.source}`,
      );
    }
  }
}

/**
 * Normalize an EnvSpec by sorting and deduplicating packages
 */
export function normalizeEnvSpec(spec: EnvSpec): EnvSpec {
  const packages = [...new Set(spec.packages)].toSorted();
  return {
    nixpkgs_pin: spec.nixpkgs_pin || DEFAULT_NIXPKGS_PIN,
    packages,
  };
}

/**
 * Compute the environment ID from a normalized EnvSpec
 * Uses sha256 of "nixpkgs_pin\npackage1\npackage2\n..."
 */
export function computeEnvId(spec: EnvSpec): string {
  const content = spec.nixpkgs_pin + "\n" + spec.packages.join("\n");
  return createHash("sha256").update(content).digest("hex").slice(0, 16);
}

/**
 * Check if an outLink symlink points to a valid store path
 */
function isValidOutLink(outLink: string): boolean {
  if (!existsSync(outLink)) {
    return false;
  }

  try {
    const stat = lstatSync(outLink);
    if (!stat.isSymbolicLink()) {
      return false;
    }

    const target = realpathSync(outLink);
    return target.startsWith("/nix/store/") && existsSync(target);
  } catch {
    return false;
  }
}

/**
 * Resolve an EnvSpec to paths and check cache status
 */
export function resolveEnv(spec: EnvSpec): ResolvedEnv {
  validateEnvSpec(spec);
  const normalized = normalizeEnvSpec(spec);
  const envId = computeEnvId(normalized);

  const specPath = join(getSpecsDir(), `${envId}.nix`);
  const outLink = join(getEnvsDir(), envId);
  const ready = isValidOutLink(outLink);

  return {
    envId,
    specPath,
    outLink,
    ready,
  };
}

File: packages/agent/agent-cli-go/internal/nixenv/resolve.go
package nixenv

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// Package name validation regex
var packageNameRegex = regexp.MustCompile(`^[a-zA-Z0-9_+\-.]+$`)

// GetAgentNixDir returns the base directory for agent-nix data
func GetAgentNixDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "share", "agent-nix")
}

// GetSpecsDir returns the directory for storing generated .nix spec files
func GetSpecsDir() string {
	return filepath.Join(GetAgentNixDir(), "specs")
}

// GetEnvsDir returns the directory for storing nix-build output symlinks
func GetEnvsDir() string {
	return filepath.Join(GetAgentNixDir(), "envs")
}

// GetAgentsDir returns the directory for per-agent runtime data
func GetAgentsDir() string {
	return filepath.Join(GetAgentNixDir(), "agents")
}

// ValidatePackageName validates a single package name
func ValidatePackageName(name string) bool {
	return packageNameRegex.MatchString(name)
}

// ValidateEnvSpec validates an EnvSpec
func ValidateEnvSpec(spec *EnvSpec) error {
	if spec.NixpkgsPin == "" {
		return fmt.Errorf("nixpkgs_pin is required")
	}

	if _, ok := NixpkgsPins[spec.NixpkgsPin]; !ok {
		allowed := make([]string, 0, len(NixpkgsPins))
		for k := range NixpkgsPins {
			allowed = append(allowed, k)
		}
		return fmt.Errorf("invalid nixpkgs_pin %q. Allowed: %s", spec.NixpkgsPin, strings.Join(allowed, ", "))
	}

	if len(spec.Packages) == 0 {
		return fmt.Errorf("packages must be a non-empty array")
	}

	for _, pkg := range spec.Packages {
		if !ValidatePackageName(pkg) {
			return fmt.Errorf("invalid package name %q", pkg)
		}
	}

	return nil
}

// NormalizeEnvSpec normalizes an EnvSpec by sorting and deduplicating packages
func NormalizeEnvSpec(spec *EnvSpec) *EnvSpec {
	// Deduplicate
	seen := make(map[string]bool)
	packages := make([]string, 0, len(spec.Packages))
	for _, pkg := range spec.Packages {
		if !seen[pkg] {
			seen[pkg] = true
			packages = append(packages, pkg)
		}
	}

	// Sort
	sort.Strings(packages)

	pin := spec.NixpkgsPin
	if pin == "" {
		pin = DefaultNixpkgsPin
	}

	return &EnvSpec{
		NixpkgsPin: pin,
		Packages:   packages,
	}
}

// ComputeEnvID computes the environment ID from a normalized EnvSpec
func ComputeEnvID(spec *EnvSpec) string {
	content := spec.NixpkgsPin + "\n" + strings.Join(spec.Packages, "\n")
	hash := sha256.Sum256([]byte(content))
	return hex.EncodeToString(hash[:])[:16]
}

// isValidOutLink checks if an outLink symlink points to a valid store path
func isValidOutLink(outLink string) bool {
	info, err := os.Lstat(outLink)
	if err != nil {
		return false
	}

	if info.Mode()&os.ModeSymlink == 0 {
		return false
	}

	target, err := filepath.EvalSymlinks(outLink)
	if err != nil {
		return false
	}

	if !strings.HasPrefix(target, "/nix/store/") {
		return false
	}

	if _, err := os.Stat(target); err != nil {
		return false
	}

	return true
}

// ResolveEnv resolves an EnvSpec to paths and checks cache status
func ResolveEnv(spec *EnvSpec) (*ResolvedEnv, error) {
	if err := ValidateEnvSpec(spec); err != nil {
		return nil, err
	}

	normalized := NormalizeEnvSpec(spec)
	envID := ComputeEnvID(normalized)

	specPath := filepath.Join(GetSpecsDir(), envID+".nix")
	outLink := filepath.Join(GetEnvsDir(), envID)
	ready := isValidOutLink(outLink)

	return &ResolvedEnv{
		EnvID:    envID,
		SpecPath: specPath,
		OutLink:  outLink,
		Ready:    ready,
	}, nil
}

// ExpandPackageSets expands set names to package lists
func ExpandPackageSets(sets []string) []string {
	packages := make([]string, 0)
	for _, set := range sets {
		set = strings.TrimSpace(set)
		if setPackages, ok := PackageSets[set]; ok {
			packages = append(packages, setPackages...)
		}
	}
	return packages
}
